### DISCUSSION TITLE
Can the MySQL and Redis passwords in the Nightingale config file be encrypted, and which encryption methods are supported?

### DISCUSSION THREAD

**Discussion by zhijieliu666** [ccfos/nightingale#1224] [Q&A] [answered]
请问一下夜莺配置文件里的mysql密码和redis密码和本身的密码支持加密吗，如果支持，支持哪种加密方式呢

**Comment by UlricQin** [comment]
最近刚有一个社区的小伙伴提了一个PR支持对配置文件的内容做加解密：https://github.com/ccfos/nightingale/commit/62867ddbf20720314ea74e19fca78290695d7a0c#diff-d215bea8bfe873ff2f65584958a6dda77462cefb5bb31a8db83e162ed0e92613 不过我还没有细研究，你可以先看看，如果需要的话可以自行编译一个版本测试下

**Reply by tanxiao1990** [reply]
晚点我写一个使用说明

**Comment by tanxiao1990** [comment, marked as answer]
> 请问一下夜莺配置文件里的mysql密码和redis密码和本身的密码支持加密吗，如果支持，支持哪种加密方式呢

1、这个功能是才提交的，当前最新的release版本5.13.1里面还没有合并进这个功能，截止（2022.10.22）当前为止，需要自行下载main分支代码打包。（或建议等下一个release版本）。

2、当前支持加密的字段

etc\server.conf 配置文件

[Redis]

Password

[DB]

DSN

[SMTP]

Pass

[Alerting.Webhook]

BasicAuthPass

[Ibex]

BasicAuthPass

[Reader]

BasicAuthPass

[[Writers]]

BasicAuthPass



etc\webapi.conf 配置文件

[Redis]

Password

[DB]

DSN

[[Clusters]]

BasicAuthPass

[Ibex]

BasicAuthPass



备注：

当然还有其他一些字段可能也需要加密，比如basicauth，JWTAuth的key字段等，第一次漏加了，后续可以加上。



3、具体使用方式：

针对不想看具体代码的使用者：

- 首先，使用明文先正常启动程序

- 然后请求webapi的以下接口进行明文加密   

  假设我想加密redis的密码字段，假设redis的密码为 123456 ，假设我使用秘钥 aaaa19900825aaaa，具体API地址为：

  http://webapihost:18000/v1/n9e/conf-prop/encrypt	

  请求参数：raw json

  {

    "key":"aaaa19900825aaaa",    //加密的秘钥，必须是16位  24位  或者 32位

    "data":"123456"          //待加密的明文

  }

  返回值：

  {

    "encrypt": "{{cipher}}o5/AyBLAqyY7bxcNqtS/xw==",  //加密后的密文

    "key": "aaaa19900825aaaa",   //加密秘钥

    "src": "123456"   // 明文

  }

  请求可以使用postman等类似工具，或者直接在服务器主机上使用curl命令进行请求。

  如果开启了[BasicAuth]，记得加入相关认证参数（-u 用户名:密码）

  假设 basicauth配置的是： user2=a6666664e5f6ghijklmnopqrstuvwxyz

  则curl命令示例如下：

  curl -u user2:a6666664e5f6ghijklmnopqrstuvwxyz -H "Content-Type: application/json" -X POST -d '{"key":"aaaa19900825aaaa","data":"123456"}' http://webapihost:18000/v1/n9e/conf-prop/encrypt

  

- 可以请求webapi的以下接口进行密文解密、反向认证加密的内容的正确性

  http://webapihost:18000/v1/n9e/conf-prop/decrypt

  请求参数： raw json

  {

    "key":"aaaa19900825aaaa",          //秘钥

    "data":"*{{cipher}}*o5/AyBLAqyY7bxcNqtS/xw=="      //待解密的密文

  }

  返回值：

  {

    "decrypt": "123456",    //解密后的明文

    "key": "aaaa19900825aaaa",   //秘钥

    "src": "{{cipher}}o5/AyBLAqyY7bxcNqtS/xw=="   //密文

  }

- 将配置文件中相关字段的值改为加密接口返回的encrypt的密文

- 修改启动程序的启动脚本，加入--key 参数，传入秘钥，重新启动程序

  nohup ./n9e server --key aaaa19900825aaaa &>> log/server.log &
  nohup ./n9e webapi --key aaaa19900825aaaa &>> log/webapi.log &

  

针对想看具体代码的使用者：

- 参考 src\pkg\secu\aes.go 文件中的 DealWithEncrypt 函数，自行写个小代码进行加密获取密文。

- 将配置文件中相关字段的值改为加密接口返回的encrypt的密文

- 修改启动程序的启动脚本，加入--key 参数，传入秘钥，重新启动程序

  nohup ./n9e server --key aaaa19900825aaaa &>> log/server.log &
  nohup ./n9e webapi --key aaaa19900825aaaa &>> log/webapi.log &

**Reply by zhijieliu666** [reply]
好的，明白了，谢谢啦，不过看这个代码里是使用AES的CBC模式加密，并没有使用随机的IV值，并且现在好像并不推荐使用AES的CBC模式，现在推荐的安全加密方法是AES的GCM模式，我再去看下要怎么改哈
